### PR TITLE
Accumulate batches in Batch-Then-Sample mode

### PR DESCRIPTION
**Pull request by MaxiBoether**
Until now, in BtS mode, we did the forward pass on a batch, downsampled it, then trained on it. This leads to different training dynamics (e.g., learning rate) because the batch size is smaller than in StB mode (or training without sampling) and is also more inefficient since we do the backward pass on smaller batches. With these changes, we accumulate the forward pass data instead and do the backward pass as soon as we reach the original batch size.

### CODE REVIEW

**Comment by MaxiBoether** [comment]
@XianzheMa

Since RHO-LOSS relies on BtS, I wanted to fix this. I hope the test shows the functionality of the accumulation. Let me know if you have any comments or see any issues with this :)

**Review by XianzheMa — CHANGES_REQUESTED** [review]
In general very good! Especially good for the added unit test now we have more confidence on the correctness of pytorch trainer!

My biggest concern is that the `train()` method is already very bloated, and now is even more bloated. Can we extract the batch accumulating logic into a separate class like:

```
class BatchAccumulator:

    def __init__(self, period):
        self.buffer = []
        # omited

    def accumulate_new_data(self, ....) -> None:
         pass
    def collect_accumulated_data(self) -> data:
        pass
    def ready_to_collect(self): bool:
        pass
```

I think this accumulation logic is not central in the `train()` method and is better to be hidden in a separate class for better readability. Hope it makes sense!

**Comment by XianzheMa** [comment]
There is another intricacy:

With batch accumulating, internally the number of training steps will not be consistent with `batch_number`, which is an index variable created by the `enumerate(train_dataloader)`. However we use `batch_number` to decide a few things for example when to checkpoint the model https://github.com/eth-easl/modyn/blob/cedada0d36fe15be06178d74a2553008f8970941/modyn/trainer_server/internal/trainer/pytorch_trainer.py#L453.

and store the `num_batches` here https://github.com/eth-easl/modyn/blob/cedada0d36fe15be06178d74a2553008f8970941/modyn/trainer_server/internal/trainer/pytorch_trainer.py#L514. (but this variable `num_batches` is not reliable anyway because it only accounts for 1 epoch.)

For experiments we make, can we just go for 50% ratio? Then we can bypass batch accumulating. I think we only need to demonstrate "selection is useful", but don't need to find out the optimal sampling ratio. Does it make sense?

**Comment by MaxiBoether** [comment]
I haven't taken a look at your other comments just yet, but I don't get your last comment fully.

1) I understand num_batches is a bit misleading here, but that's a fixable issue (and should not affect training, just some meta stuff like checkpointing)

2) I don't understand what you mean by 50% ratio for the experiments? 50% subset selection? How do you get to that? In general, yes, for now we are mainly looking for _any_ scenario where we find improvements, let it be 10%, 30%, 50% or whatnot. But in any case, we should properly accumulate batches to have the same training dynamics for the learning rate, batch size etc - I am happy to solve the issues you mention, I might not see what is the blocker here

**Comment by XianzheMa** [comment]
> 2\. atches

Yes you are right! I think I confused myself :)

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 1]
```diff
@@ -370,6 +370,21 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
...
+            expected_bts_size = max(int(self._downsampler.downsampling_ratio * self._batch_size / 100), 1)
+            assert expected_bts_size < self._batch_size
+            if self._batch_size % expected_bts_size != 0:
+                raise ValueError(
+                    f"The target batch size of {self._batch_size} is not a multiple of the batch size "
+                    + f"after downsampling a batch in BtS mode ({expected_bts_size}). We cannot accumulate batches. "
+                    + "Please choose the downsampling ratio and batch size such that this is possible."
+                )
+
+            bts_accumulate_period = int(self._batch_size / expected_bts_size)
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:386]: I will just go for // since we have the guarantee that self._batch_size % expected_bts_size == 0 here.

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 2]
```diff
@@ -418,6 +433,66 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
...
+                                if isinstance(data, dict)
+                                else data.detach().cpu()
+                            )
+
+                            bts_accumulation_buffer.append(
+                                (data, sample_ids, target.detach().cpu(), weights.detach().cpu())
+                            )
+                            stopw.start("FetchBatch", resume=True)
+                            stopw.start("IndivFetchBatch", overwrite=True)
+                            continue
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:480]: I think this is ok - if we imagine training without BtS and batch size x, we only consider full batches (IIRC, the pytorch dataloader drops incomplete batches). This means we only consider a number of batches that is divisible

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 3]
```diff
@@ -418,6 +433,66 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
...
+                            f"expected bts size: {expected_bts_size} "
+                            + f"actual: {data.shape[0] if isinstance(data, torch.Tensor) else 'n/a'}"
+                        )
+
+                        assert (
+                            len(sample_ids) == expected_bts_size
+                        ), f"expected bts size: {expected_bts_size} actual: {len(sample_ids)}"
+                        assert (
+                            target.shape[0] == expected_bts_size
+                        ), f"expected bts size: {expected_bts_size} actual: {target.shape[0]}"
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:450]: I will move the assertions to a separate function to debloat `train`, but we need them because the output size here depends on the downsamplers outputting the expected number of items. Since the trainer cannot force them to, I'd rather assert this here

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 4]
```diff
@@ -370,6 +370,21 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
             # an infinity epoch generator
             epoch_num_generator = itertools.count(start=0)
             self._info(f"Training will stop when the number of samples to pass reaches {self.num_samples_to_pass}.")
+
+        bts_accumulation_buffer: list[tuple] = []
+        if self._downsampling_mode == DownsamplingMode.BATCH_THEN_SAMPLE:
+            # We cannot pass the target size from the trainer server since that depends on StB vs BtS.
+            expected_bts_size = max(int(self._downsampler.downsampling_ratio * self._batch_size / 100), 1)
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:377]: It's what we expect from the downsampler after downsampling, hence I chose that name. I will go for `post_downsampling_size` for now (but we lose the information that this is what we expect - we later on `assert` that this actually happens)

File: modyn/tests/trainer_server/internal/trainer/test_pytorch_trainer.py [hunk 5]
```diff
@@ -824,3 +844,100 @@ def test_create_trainer_with_exception(
...
+
+    # Mock the downsample_batch method to return batches of the expected size
+    expected_bts_size = int(batch_size * (downsampling_ratio / 100.0))
+    bts_accumulate_period = batch_size // expected_bts_size
+
+    global num_downsamples
+    num_downsamples = -1
+
+    def mock_downsample_batch(data, sample_ids, target):
+        global num_downsamples
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/test_pytorch_trainer.py:903]: Thanks, that's a lot cleaner. Adjusted.

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 6]
```diff
@@ -418,6 +433,66 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
                         stopw.start("DownsampleBTS", resume=True)
                         data, sample_ids, target, weights = self.downsample_batch(data, sample_ids, target)
```
[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:434]: Adjusted

**Comment by MaxiBoether** [comment]
Thanks for your feedback @XianzheMa, the `train` function looks neater again now :)

File: modyn/trainer_server/internal/trainer/batch_accumulator.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        Retrieves the accumulated batch of data.
+
+        Returns:
+            tuple[torch.Tensor | dict[str, torch.Tensor], list, torch.Tensor, torch.Tensor]:
+                - The accumulated input data.
+                - The accumulated sample IDs.
+                - The accumulated target values.
+                - The accumulated weights.
+        """
+        data, sample_ids, target, weights = self._accumulation_buffer[-1]
```
Comment (XianzheMa) [inline modyn/trainer_server/internal/trainer/batch_accumulator.py:63]: Shouldn't we first move data, target, weights to the target device?

Otherwise I think we cannot concatenate A and B where A is on cpu and B is moved to gpu.

File: modyn/trainer_server/internal/trainer/batch_accumulator.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            target (torch.Tensor): The target values for the batch.
+            weights (torch.Tensor): The weights for the batch.
+
+        Returns:
+            bool: True if the accumulation period is reached and the accumulated batch is ready, False otherwise.
+        """
+        data = {k: v.detach().cpu() for k, v in data.items()} if isinstance(data, dict) else data.detach().cpu()
+
+        self._accumulation_buffer.append((data, sample_ids.copy(), target.detach().cpu(), weights.detach().cpu()))
+
```
Comment (XianzheMa) [inline modyn/trainer_server/internal/trainer/batch_accumulator.py:44]: can be simplified as
```python
self._current_batch_number = (self._current_batch_number + 1) % self._accumulation_period
return self._current_batch_number == 0
```

File: modyn/tests/trainer_server/internal/trainer/test_batch_accumulator.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        self.assertIsInstance(accumulated_data, torch.Tensor)
+        self.assertEqual(accumulated_data.shape, (6, 3))
+        self.assertEqual(len(accumulated_sample_ids), 6)
+        self.assertEqual(accumulated_target.shape, (6,))
+        self.assertEqual(accumulated_weights.shape, (6,))
+        self.assertEqual(len(self.accumulator._accumulation_buffer), 0)
+
+        assert torch.allclose(
+            accumulated_data, torch.Tensor([[12, 13, 14], [15, 16, 17], [6, 7, 8], [9, 10, 11], [1, 2, 3], [3, 4, 5]])
+        )
```
Comment (XianzheMa) [inline modyn/tests/trainer_server/internal/trainer/test_batch_accumulator.py:124]: Why do we need to reverse the oder of the data? (I should ask this question in the implementation but :D)

I thought it is natural to have `[[1, 2, 3], [3, 4, 5], [6, 7, 8], [9, 10, 11], [12, 13, 14], [15, 16, 17]]` as the result.

**Review by XianzheMa — APPROVED** [review]
Only two questions on the reversed order and moving the data to the target device. Besides this LGTM!
Once I have implemented RHO-LOSS, I will create an integration test covering both on RHO-LOSS and this.

File: modyn/trainer_server/internal/trainer/batch_accumulator.py (hunk 7 above)
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/batch_accumulator.py:63]: Thanks, fixed together with the reverse accumulation

File: modyn/trainer_server/internal/trainer/batch_accumulator.py (hunk 8 above)
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/batch_accumulator.py:44]: Thanks!

File: modyn/tests/trainer_server/internal/trainer/test_batch_accumulator.py (hunk 9 above)
Comment (MaxiBoether) [reply modyn/tests/trainer_server/internal/trainer/test_batch_accumulator.py:134]: Historic reasons with the old implementation which made this easier. Fixed now

**Comment by MaxiBoether** [comment]
Thanks @XianzheMa. When you resolve your comments and CI runs through, we can merge this :)

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 10]
```diff
@@ -534,6 +556,21 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
...
 
+    @staticmethod
+    def _assert_data_size(
+        expected_size: int, data: Union[torch.Tensor, dict[Any, torch.Tensor]], sample_ids: list, target: torch.Tensor
+    ) -> None:
+        assert (
+            all(tensor.shape[0] == expected_size for tensor in data.values())
+            if isinstance(data, dict)
+            else data.shape[0] == expected_size
+        ), (
```
[the remark this answers is not in the document]
Comment (XianzheMa) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:567]: Is it a failure from a unit test, integration test,  or your experiment?

[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:567]: Hm, thinking about it, we do not set drop_last=True on the dataloader. In a config that fails, could you try setting drop last to True and see if that fixed the issue?

[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:567]: > Is it a failure from a unit test, integration test, or your experiment?

Unit tests and integration tests are passing on main. It's my experiment and also other pipeline configurations that reside within the repo.

[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:567]: > Hm, thinking about it, we do not set drop_last=True on the dataloader. In a config that fails, could you try setting drop last to True and see if that fixed the issue?

maybe it was renamed, but i cannot find such a `drop_last` option, also not in older commits.

[the remark this answers is not in the document]
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:567]: No, I mean the drop last option of the torch DataLoader: https://pytorch.org/docs/stable/data.html

that should be in instantiate dataloader or sth like this, in the data utils

[the remark this answers is not in the document]
Comment (robinholzi) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:567]: @MaxiBoether Thanks! That seems to do the trick. Are you confident that setting this flag whenever `batch_size` is present (which is always in our case) will not have any significant effect on training performance? Not sure if ignoring the last batch makes sense in all instances.
